TombEngine/Objects/Effects/Boss.cpp

```diff
@@ -0,0 +1,318 @@
+#include "framework.h"
+#include "Objects/Effects/Boss.h"
+
+#include "Game/collision/collide_room.h"
+#include "Game/effects/effects.h"
+#include "Game/effects/spark.h"
+#include "Game/items.h"
+#include "Game/misc.h"
+#include "Objects/TR3/Entity/PunaBoss.h"
+#include "Specific/setup.h"
+
+using namespace TEN::Effects::Spark;
+using namespace TEN::Entities::Creatures::TR3;
+
+namespace TEN::Effects::Boss
+{
+	void SpawnShield(const ItemInfo& item, const Vector4& color)
+	{
+		if (!item.TestFlags((int)BossItemFlags::Object, (short)BossFlagValue::Shield))
+			return;
+
+		int itemNumber = CreateItem();
+		if (itemNumber == NO_ITEM)
+			return;
+
+		auto& shieldItem = g_Level.Items[itemNumber];
+
+		shieldItem.ObjectNumber = ID_BOSS_SHIELD;
+		shieldItem.RoomNumber = item.RoomNumber;
+		shieldItem.Pose.Position = item.Pose.Position;
+		shieldItem.Pose.Position.y -= CLICK(3);
+		shieldItem.Pose.Orientation = item.Pose.Orientation;
+		shieldItem.Flags |= IFLAG_ACTIVATION_MASK;
+
+		InitialiseItem(itemNumber);
+
+		shieldItem.Model.Color = color;
+		shieldItem.Collidable = true;
+		shieldItem.ItemFlags[0] = 2;
+		shieldItem.Model.Mutator[0].Scale = Vector3(2.0f); // Scale model by factor of 2.
+		shieldItem.Status = ITEM_ACTIVE;
+
+		AddActiveItem(itemNumber);
+	}
+
+	void SpawnShockwaveExplosion(const ItemInfo& item, const Vector4& color)
+	{
+		if (!item.TestFlags((int)BossItemFlags::Object, (short)BossFlagValue::ShockwaveExplosion))
+			return;
+
+		int itemNumber = CreateItem();
+		if (itemNumber == NO_ITEM)
+			return;
+
+		auto& shockwaveItem = g_Level.Items[itemNumber];
+
+		shockwaveItem.Pose.Position = item.Pose.Position;
+		shockwaveItem.Pose.Position.y -= CLICK(2);
+
+		shockwaveItem.Pose.Orientation = EulerAngles(
+			Random::GenerateAngle(ANGLE(-20.0f), ANGLE(20.0f)),
+			Random::GenerateAngle(ANGLE(-20.0f), ANGLE(20.0f)),
+			Random::GenerateAngle(ANGLE(-20.0f), ANGLE(20.0f)));
+
+		shockwaveItem.ObjectNumber = ID_BOSS_EXPLOSION_SHOCKWAVE;
+		shockwaveItem.RoomNumber = item.RoomNumber;
+		shockwaveItem.Flags |= IFLAG_ACTIVATION_MASK;
+
+		InitialiseItem(itemNumber);
+
+		auto result = color;
+		result.w = 1.0f;
+		shockwaveItem.Model.Color = result;
+		shockwaveItem.Collidable = false;					 // No collision for this entity.
+		shockwaveItem.ItemFlags[0] = 70;						 // Timer before clearing; will fade out, then get destroyed.
+		shockwaveItem.Model.Mutator[0].Scale = Vector3::Zero; // Start without scale.
+		shockwaveItem.Status = ITEM_ACTIVE;
+
+		AddActiveItem(itemNumber);
+		SoundEffect(SFX_TR3_BLAST_CIRCLE, &shockwaveItem.Pose);
+	}
+
+	void SpawnShockwaveRing(const ItemInfo& item, const Vector3& pos, const Vector4& color)
```

Comment:
Is it possible to use existing TR4 shockwave effect instead of using dummy object mesh?

Reply:
Done